src/main/kotlin/ecommerce/service/ProductService.kt

```diff
@@ -0,0 +1,35 @@
+package ecommerce.service
+
+import ecommerce.model.Product
+import ecommerce.model.ProductPatchDTO
+import ecommerce.store.ProductStore
+import org.springframework.stereotype.Service
+
+@Service
+class ProductService(private val productStore: ProductStore) {
+    fun findAll(): List<Product> = productStore.findAllProducts()
+
+    fun findById(id: Long): Product? = productStore.findProductById(id)
+
+    fun insert(product: Product): Product = productStore.insertProduct(product)
+
+    fun update(
+        id: Long,
+        dto: ProductPatchDTO,
+    ): Boolean {
+        val product = productStore.findProductById(id) ?: return false
+
+        dto.name?.let { product.name = it }
+        dto.price?.let { product.price = it }
+        dto.imageUrl?.let { product.imageUrl = it }
```

Comment:
```suggestion
        with(dto) {
            apply {
                product.name = name ?: product.name
                product.price = price ?: product.price
                product.imageUrl = imageUrl ?: product.imageUrl
            }
        }
```

You could rewrite the code like this using `with` and `apply` 👀

Reply:
```kotlin
product.name = name ?: product.name
product.price = price ?: product.price
product.imageUrl = imageUrl ?: product.imageUrl
```

I also think it would be better to delegate this logic to `Product`.

Reply:
Moved to Product class. Now I guess that will simplify the validation logic. Thanks!